pulser/simulation/simulation.py

```diff
@@ -127,16 +135,17 @@ def draw(self, draw_phase_area=False):
             self._seq, self.sampling_rate, draw_phase_area=draw_phase_area
         )
 
-    def _extract_samples(self):
+    def _extract_samples(self) -> None:
         """Populate samples dictionary with every pulse in the sequence."""
 
-        def prepare_dict():
+        def prepare_dict() -> dict[str, np.ndarray]:
             # Duration includes retargeting, delays, etc.
             return {'amp': np.zeros(self._tot_duration),
                     'det': np.zeros(self._tot_duration),
                     'phase': np.zeros(self._tot_duration)}
 
-        def write_samples(slot, samples_dict):
+        def write_samples(slot: _TimeSlot,
+                          samples_dict: Mapping[str, np.ndarray]) -> None:
             samples_dict['amp'][slot.ti:slot.tf] += slot.type.amplitude.samples
             samples_dict['det'][slot.ti:slot.tf] += slot.type.detuning.samples
             samples_dict['phase'][slot.ti:slot.tf] = slot.type.phase
```

Comment:
You need to cast `slot.type` to `Pulse`. Do it just once, before these lines. You can add something like: 
```
pulse = cast(Pulse, slot.type)
```
and then use `pulse` instead of `slot.type` on these lines.